third_party/move/move-binary-format/src/file_format.rs

```diff
@@ -1501,21 +1524,119 @@ impl SignatureToken {
             Address => Address,
             Signer => Signer,
             Vector(ty) => Vector(Box::new(ty.instantiate(subst_mapping))),
+            Function(args, result, abilities) => {
+                Function(inst_vec(args), inst_vec(result), *abilities)
+            },
             Struct(idx) => Struct(*idx),
-            StructInstantiation(idx, struct_type_args) => StructInstantiation(
-                *idx,
-                struct_type_args
-                    .iter()
-                    .map(|ty| ty.instantiate(subst_mapping))
-                    .collect(),
-            ),
+            StructInstantiation(idx, struct_type_args) => {
+                StructInstantiation(*idx, inst_vec(struct_type_args))
+            },
             Reference(ty) => Reference(Box::new(ty.instantiate(subst_mapping))),
             MutableReference(ty) => MutableReference(Box::new(ty.instantiate(subst_mapping))),
             TypeParameter(idx) => subst_mapping[*idx as usize].clone(),
         }
     }
 }
 
+/// A `ClosureMask` is a value which determines how to distinguish those function arguments
+/// which are captured and which are not when a closure is constructed. For instance,
+/// with `_` representing an omitted argument, the mask for `f(a,_,b,_)` would have the argument
+/// at index 0 and at index 2 captured. The mask can be used to transform lists of types.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+#[cfg_attr(any(test, feature = "fuzzing"), derive(proptest_derive::Arbitrary))]
+#[cfg_attr(any(test, feature = "fuzzing"), proptest(no_params))]
+#[cfg_attr(
+    feature = "fuzzing",
+    derive(arbitrary::Arbitrary),
+    derive(dearbitrary::Dearbitrary)
+)]
+pub struct ClosureMask {
+    pub mask: u64,
+}
+
+impl fmt::Display for ClosureMask {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:b}", self.mask)
+    }
+}
+
+impl ClosureMask {
+    pub fn new(mask: u64) -> Self {
+        Self { mask }
+    }
+
+    /// Apply a closure mask to a list of elements, returning only those
+    /// where position `i` is set in the mask (if `collect_captured` is true) or not
+    /// set (otherwise).
+    pub fn extract<T: Clone>(&self, tys: &[T], collect_captured: bool) -> Vec<T> {
+        tys.iter()
+            .enumerate()
+            .filter_map(|(pos, x)| {
+                let set = (1 << pos) & self.mask != 0;
```

Comment:
Seems like a bug, if `tys.len() > 64` shift will panic. We should enforce here the number of types is <= 64. I have not seen (or maybe missed) where we enforce this in bytecode verifier.